TechnicalDocuments/Proposals/TopPriorityMemberVariables.md

```diff
@@ -0,0 +1,144 @@
+# Top-priority member-variable and initializer declarations
+
+* Author: Yasuhiro Inami
+* Review Manager: Rui Peres
+
+## Introduction
+
+This proposal is to improve code readability for every declared type by **moving member variables (states) and initializers (constructors) to the top of the scope**.
+
+## Motivation
+
+The code becomes hard to read when types are heavily nested, especially when inner type's member variables appear before outer type's variables in the file.
+
+For example, below code shows that we can't immediately figure out which member variables belong to which type, and what kind of constructors are available:
+
+```swift
+enum Outer {
+    struct Inner {
+        func foo() {
+            ...
+            ...
+            ...
+        }
+
+        let a: A
+        let b: B
+
+        init(a: A, b: B) {
+            ...
+        }
+    }
+
+    func foo() {
+        ...
+    }
+
+    case c
+
+    var bar: Bar {
+        ...
+    }
+
+    init?(c: C) {
+        ...
+    }
+
+    case d(D)
+}
+```
+
+Since understanding the data structure i.e. product type (member variables a.k.a. states) and sum type (enum cases a.k.a. constructors) are the key to understand the types, architecture, and whole business logic, it is important to organize the code as proposed in the next section.
+
+## Proposed solution
+
+**Move member variables (states) and initializers (constructors) to the top of the scope**, at least manually, and in future, automatically by formatter tool e.g. SwiftSyntax.
+
+For example:
+
+```swift
+// Group constructors (enum cases).
+enum Outer {
+    case c
+    case d(D)
+
+    // NOTE: This can be implemented as `extension Outer` as well
+    init?(c: C) {
+        ...
+    }
+}
+
+// Methods (non-constructors) should move to extension scope to separate from constructors.
+// Or, they can be in the same scope as `enum Outer` but MUST NOT be at the top of its scope.
```

Comment:
> Methods (non-constructors) should move to extension scope to separate from constructors.

I'd stick with having them in the same scope unless there's a reason to group them together, e.g. implementing a protocol.